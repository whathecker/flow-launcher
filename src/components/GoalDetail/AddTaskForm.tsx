/* eslint-disable no-console */
import React, { useContext } from "react";
import { GoalsContext } from "../../contexts/goals";
import { TextInput, StyleSheet } from "react-native";
import { View } from "../Themed";
import { Button } from "../shared";

import { Container, Typography, Color } from "../../styles";

import { IAddTaskInput } from "../../types/core/entity";
import { Formik } from "formik";
import * as Yup from "yup";

type AddTaskFormProps = {
  goal_id: string;
  addTaskFormOpenedHandler: React.Dispatch<React.SetStateAction<boolean>>;
};

const ValidationSchema = Yup.object().shape({
  title: Yup.string().required("Required"),
  description: Yup.string(),
});

function isFormReadyToSubmit(values: IAddTaskInput): boolean {
  return values.title.length > 0;
}

const AddTaskForm: React.FC<AddTaskFormProps> = ({
  goal_id,
  addTaskFormOpenedHandler,
}: AddTaskFormProps) => {
  const { addTaskToGoal } = useContext(GoalsContext);
  return (
    <Formik
      initialValues={{ title: "", description: "", goal_id } as IAddTaskInput}
      validationSchema={ValidationSchema}
      onSubmit={async (values) => {
        try {
          await addTaskToGoal(values);
          addTaskFormOpenedHandler(false);
        } catch (error) {
          // TOOD: add error handling
          console.error(error);
        }
      }}
    >
      {({ handleChange, handleSubmit, values }) => {
        return (
          <View>
            <View style={styles.titleInputWrapper}>
              <TextInput
                style={styles.titleInput}
                multiline
                scrollEnabled
                autoFocus={true}
                onChangeText={handleChange("title")}
                placeholder={"Write your task in a sentence"}
                placeholderTextColor={Color.light.subtleLabel}
                value={values.title}
              />
            </View>
            <View style={styles.descInputWrapper}>
              <TextInput
                style={styles.descInput}
                multiline
                scrollEnabled
                onChangeText={handleChange("description")}
                placeholder={"Write the description of your task"}
                placeholderTextColor={Color.light.subtleLabel}
                value={values.description}
              />
            </View>
            <View style={styles.buttonAreaWrapper}>
              <View style={styles.buttonWrapper}>
                <Button
                  disable={isFormReadyToSubmit(values) ? false : true}
                  ctaTxt="Add"
                  pressHandler={handleSubmit}
                />
              </View>
            </View>
          </View>
        );
      }}
    </Formik>
  );
};

const styles = StyleSheet.create({
  titleInputWrapper: {
    width: "95%",
    paddingLeft: 20,
    paddingTop: 30,
    height: "28%",
    maxHeight: "28%",
  },
  titleInput: {
    ...Typography.p,
    fontSize: 22,
  },
  descInputWrapper: {
    width: "95%",
    paddingLeft: 20,
    paddingBottom: 20,
    height: "50%",
    maxHeight: "50%",
  },
  descInput: {
    ...Typography.p,
    fontSize: 16,
    //minHeight: "100%",
  },
  buttonAreaWrapper: {
    ...Container.flexStart,
    height: "18%",
    justifyContent: "flex-end",
    marginTop: "1.5%",
    marginRight: 5,
  },
  buttonWrapper: {
    width: "50%",
  },
});

export default AddTaskForm;
