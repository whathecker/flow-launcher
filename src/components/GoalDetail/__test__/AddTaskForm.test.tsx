import * as React from "react";
import renderer from "react-test-renderer";
import AddTaskForm from "../AddTaskForm";

describe("Test AddTaskForm component", () => {
  it("Should renders AddTaskForm correctly", () => {
    const mockSetState = () => false;
    const tree = renderer
      .create(
        <AddTaskForm
          goal_id="mock_goal"
          addTaskFormOpenedHandler={mockSetState}
        />,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
