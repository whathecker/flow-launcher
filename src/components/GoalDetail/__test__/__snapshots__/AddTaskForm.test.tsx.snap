// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test AddTaskForm component Should renders AddTaskForm correctly 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#fff",
      },
      undefined,
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#fff",
        },
        Object {
          "height": "20%",
          "paddingLeft": 20,
          "paddingTop": 30,
        },
      ]
    }
  >
    <TextInput
      allowFontScaling={true}
      autoFocus={true}
      multiline={true}
      onChangeText={[Function]}
      placeholder="Write your task in a sentence"
      placeholderTextColor="#848484"
      rejectResponderTermination={true}
      style={
        Object {
          "fontFamily": "Lexend-Regular",
          "fontSize": 22,
        }
      }
      underlineColorAndroid="transparent"
      value=""
    />
  </View>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#fff",
        },
        Object {
          "height": "60.5%",
          "paddingLeft": 20,
          "paddingTop": 20,
        },
      ]
    }
  >
    <TextInput
      allowFontScaling={true}
      multiline={true}
      onChangeText={[Function]}
      placeholder="Write the description of your task"
      placeholderTextColor="#848484"
      rejectResponderTermination={true}
      style={
        Object {
          "fontFamily": "Lexend-Regular",
          "fontSize": 16,
          "height": "125%",
        }
      }
      underlineColorAndroid="transparent"
      value=""
    />
  </View>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#fff",
        },
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "flex-end",
          "marginRight": 5,
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#fff",
          },
          Object {
            "width": "50%",
          },
        ]
      }
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#848785",
            "borderRadius": 10,
            "flexDirection": "row",
            "justifyContent": "center",
            "opacity": 1,
            "paddingBottom": 20,
            "paddingTop": 20,
            "shadowColor": "#171717",
            "shadowOffset": Object {
              "height": 5,
              "width": 3,
            },
            "shadowOpacity": 0.2,
            "shadowRadius": 3,
            "width": "100%",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#fff",
              },
              Object {
                "fontFamily": "Lexend-Regular",
                "fontSize": 18,
              },
            ]
          }
        >
          Add
        </Text>
      </View>
    </View>
  </View>
</View>
`;
